src/pam/main.cc

```diff
@@ -0,0 +1,394 @@
+#include <condition_variable>
+#include <cstring>
+#include <fstream>
+#include <functional>
+#include <future>
+#include <iostream>
+#include <iterator>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <system_error>
+#include <thread>
+#include <tuple>
+#include <vector>
+
+#include <INIReader.h>
+
+#include <boost/locale.hpp>
+
+#include <security/pam_appl.h>
+#include <security/pam_ext.h>
+#include <security/pam_modules.h>
+
+#include "main.hh"
+#include "optional_task.hh"
+
+using namespace std;
+using namespace boost::locale;
+using namespace std::chrono_literals;
+
+/**
+ * Inspect the status code returned by the compare process
+ * @param  code          The status code
+ * @param  conv_function The PAM conversation function
+ * @return               A PAM return code
+ */
+int on_howdy_auth(int code, function<int(int, const char *)> conv_function) {
+  // If the process has exited
+  if (!WIFEXITED(code)) {
+    // Get the status code returned
+    code = WEXITSTATUS(code);
+
+    switch (code) {
+    // Status 10 means we couldn't find any face models
+    case 10:
+      conv_function(PAM_ERROR_MSG,
+                    dgettext("pam", "There is no face model known"));
+      syslog(LOG_NOTICE, "Failure, no face model known");
+      break;
+    // Status 11 means we exceded the maximum retry count
+    case 11:
+      syslog(LOG_INFO, "Failure, timeout reached");
+      break;
+    // Status 12 means we aborted
+    case 12:
+      syslog(LOG_INFO, "Failure, general abort");
+      break;
+    // Status 13 means the image was too dark
+    case 13:
+      conv_function(PAM_ERROR_MSG,
+                    dgettext("pam", "Face detection image too dark"));
+      syslog(LOG_INFO, "Failure, image too dark");
+      break;
+    // Otherwise, we can't describe what happened but it wasn't successful
+    default:
+      conv_function(
+          PAM_ERROR_MSG,
+          string(dgettext("pam", "Unknown error:") + to_string(code)).c_str());
+      syslog(LOG_INFO, "Failure, unknown error %d", code);
+    }
+  }
+
+  // As this function is only called for error status codes, signal an error to
+  // PAM
+  return PAM_AUTH_ERR;
+}
+
+/**
+ * Format and send a message to PAM
+ * @param  conv    PAM conversation function
+ * @param  type    Type of PAM message
+ * @param  message String to show the user
+ * @return         Returns the conversation function return code
+ */
+int send_message(function<int(int, const struct pam_message **,
+                              struct pam_response **, void *)>
+                     conv,
+                 int type, const char *message) {
+  // No need to free this, it's allocated on the stack
+  const struct pam_message msg = {.msg_style = type, .msg = message};
+  const struct pam_message *msgp = &msg;
+
+  struct pam_response res_ = {};
+  struct pam_response *resp_ = &res_;
+
+  // Call the conversation function with the constructed arguments
+  return conv(1, &msgp, &resp_, nullptr);
```

Comment:
> Hey everyone. I've been using howdy for a while and thought I'd try this version. Does anyone have this working with KDE? For me, it does not work with the KDE screen locker (kscreenlocker_greet) at all. In fact even my password gets denied when this is enabled.
> 
> The other thing I noticed is that when it's used with sudo, terminal input is not visible until I issue "stty echo". Although, it's probably not secure to use face unlock with sudo anyways.
> 
> Not sure if I'm smart enough to figure it out, but I'm happy to test. Unlocking is much faster with this new version so I'm excited about the future of this project.

Figured this out. `kcheckpass`, which is responsible for doing the auth with PAM on behalf of the KDE screenlocker, is getting a segfault in the conversation function: https://github.com/KDE/kscreenlocker/blob/084e6ab3b700b0354ec8d85d2e690773b51782b4/kcheckpass/checkpass_pam.c#L53.

The reason is that the pam module is supposed to send a user-defined `appdata_ptr` as the last argument to the `conv` function, but in our case, it is set to `nullptr` here.

All that needs to be done is to extract this `appdata_ptr` from the pam `conv` struct and pass it to this callback. This solves the crash.

Reply:
Thanks for the digging! I will patch it ASAP.